source/nodejs/adaptivecards-fabric/src/Components/Inputs/InputToggleFabric.tsx

```diff
@@ -14,6 +14,7 @@ export class InputToggleFabric extends Shared.ReactInputElement {
     public parse = (json: any, errors?: AC.IValidationError[]) => {
         this.id = AC.getStringValue(json.id);
         this.valueInternal = json.value;
+        this.defaultValue = this.valueInternal;
```

Comment:
There are a number of things that need to be changed - I am sorry I didn't spot them earlier, I reviewed this too fast:
- Your custom inputs extend the AC.Input class. Your `parse` method **MUST** first call `super.parse()` which already parses `id` and `defaultValue`. Please don't reimplement these behaviors.
- In your `internalRender` implementation, you have to initialize the value of your actual Fabric input control to the `defaultValue` property
- Your current internalRender implementation does this: `this.inputControlContainerElement.style.display = "block"` **Don't do that**; you mustn't change the layout of any element you do not own (e.g create yourself) otherwise you will interfere with the layout of the card.
- Your inputs _should_ also implement `toJSON()` (please look at how it's implemented for one of the built-in inputs). I say they should because technically toJSON is only used by the AC designer which doesn't use the Fabric extensions.

Reply:
@dclaux, trying to understand a bit more -- are these guidelines in place for some of these do's/don't's you point out? I am assuming most of these points are more guidelines and wouldn't actually cause functionality/runtime errors? Or would they? (trying to see if we need CI's in place or some other automation that can catch some of these violations eventually)

Reply:
imo, would be really nice to have these requirements documented somewhere. I think only **style.display** potentially can cause some rendering issues, others are mostly guidelines.

Reply:
@shalinijoshi19 these are more than guidelines, this is how to correctly extend the renderer. When we add new features, such as validation on inputs, or new properties in general, we don't want nor can we update everyone's custom input implementation. So the renderer provides the core behavior and extensions just need to implement the value add they're bringing, in this case a better looking, more functional control.

Extending the renderer should be better documented. It is unfortunate that it is not at this point, but the least I can do right now is point folks in the right direction.

Reply:
@dfisenko the general rule is do not modify any element not created by you. That is especially true when it comes to layout, but it is also true for styles. You can however set the layout and styles of your custom element whichever way you desire/makes sense.

The main role of the renderer is to take care of layout and styles as defined in the Host Configuration. Overriding the styles of any element that was created under you can result in the renderer not rendering properly. While that is ok for as long as the extensions are yours and your are the only one using them, in this case the extensions are meant to be used by other developers.

I hope that makes sense.

Technically the proper way to extend the renderer is already documented here: https://docs.microsoft.com/en-us/adaptive-cards/sdk/rendering-cards/javascript/extensibility
That being said, this documentation could probably be more extensive.